In [1]:
!ls

artist_genres.csv		cifras_scraper_container.tar.gz
cifras				cifras_scraper.log
cifras2				cifras_scraper.py
cifras2.csv			cifra-web-scrape-study.ipynb
cifras3.csv			container
cifras4.csv			genre_scraper.py
cifras.csv			genres_scraper.log
cifras_data_analysis.ipynb	initial-data-analysis.ipynb
cifras-data-availability.ipynb	initial-full-data-analysis.ipynb
cifras_df2.csv			README.md
cifras_scraper2.log		sitemap
cifras_scraper2.py		sitemap_index.xml
cifras_scraper3.log		sitemap_index.xml.gz
cifras_scraper3_old.log		songs_list_br.ipynb
cifras_scraper3.py		spotify-data-scrape.ipynb
cifras_scraper4.log


# Data Structure Consolidation

- During the scraping period, we change the data acquired during the process for multiple reasons. Files ```cifras_scraper.py```,```cifras_scraper2.py```,```cifras_scraper3.py```, and ```cifras_scraper4.py``` were used to scrape data from the website and acquire cifras and information about them. There is also a respective log file for each Python cifra scrape file. Since the data was stored using different codes, it is necessary to consolidate all data acquired in a single structure.
- Cifras are stored in folders ```cifras/``` and ```cifras2/```, while information regarding each cifra is stored in csv files: ```cifras.csv```, ```cifras2.csv```, ```cifras3.csv```, and ```cifras4.csv```.
- Also, there is information about artists' music genre in the file ```artist_genres.csv```, which was obtained via ```genre_scraper.py``` with its respective log file.

**Let's use pandas library to check the data available in each csv file!**

In [2]:
import pandas as pd

cifras_df1 = pd.read_csv("cifras.csv")
cifras_df2 = pd.read_csv("cifras2.csv")
cifras_df3 = pd.read_csv("cifras3.csv")
cifras_df4 = pd.read_csv("cifras4.csv")
artist_genre = pd.read_csv("artist_genres.csv")

In [3]:
cifras_df1.describe(include=[object])

url_artist        url_song  \
count              4063            4063   
unique             2501            3870   
top     banda universos  espirito santo   
freq                 52               6   

                                     cifra_file_loc cifra_key  \
count                                          4063      3748   
unique                                         4063        32   
top     cifras/gesil-amarante-jr/sou-um-milagre.txt         G   
freq                                              1       517   

         spotify_artist    spotify_song   spotify_album release_date  \
count              4063            4063            4063         4063   
unique             2706            3951            3343         1727   
top     Banda Universos  Espírito Santo  Tempo para Mim   2023-05-26   
freq                 50               5              12           28   

                                                cifra_url  \
count                                                4063   
unique                                               4063   
top     https://www.cifras.com.br/cifra/gesil-amarante...   
freq                                                    1   

                                              spotify_url  
count                                                4063  
unique                                               4053  
top     https://open.spotify.com/track/1Vcx2i6hnLdxJJU...  
freq                                                    3

In [4]:
cifras_df1.head()

url_artist                 url_song  \
0  gesil amarante jr           sou um milagre   
1         may soares         nao passes daqui   
2      luciana gomes  consequencia do sucesso   
3   gabriel iglesias               velho amor   
4        indiegentes     eu nao sou teu amigo   

                                     cifra_file_loc cifra_key  \
0       cifras/gesil-amarante-jr/sou-um-milagre.txt         F   
1            cifras/may-soares/nao-passes-daqui.txt         A   
2  cifras/luciana-gomes/consequencia-do-sucesso.txt         A   
3            cifras/gabriel-iglesias/velho-amor.txt         A   
4       cifras/indiegentes/eu-nao-sou-teu-amigo.txt        Em   

      spotify_artist             spotify_song            spotify_album  \
0  Gesil Amarante Jr           Sou um Milagre           Sou um Milagre   
1         May Soares         Não Passes Daqui         Não Passes Daqui   
2      LUCIANA GOMES  Consequência do Sucesso  Consequência do Sucesso   
3   Gabriel Iglesias               Velho Amor           Pedra em Carne   
4        Indiegentes     Eu Não Sou Teu Amigo     Eu Não Sou Teu Amigo   

   spotify_popularity release_date  feature_danceability  ...  feature_key  \
0                   2   2023-09-21                 0.425  ...            5   
1                   3   2023-08-12                 0.414  ...            9   
2                   0   2022-07-19                 0.621  ...            7   
3                   7   2016-10-14                 0.453  ...            9   
4                   0   2018-12-07                 0.182  ...            4   

   feature_liveness  feature_mode  feature_speechiness  feature_tempo  \
0            0.1240             1               0.0322        164.912   
1            0.2020             1               0.0276        140.875   
2            0.2260             1               0.0336        139.987   
3            0.0903             1               0.0335         92.835   
4            0.1020             0               0.0335        172.774   

   feature_time_signature  feature_valence  song_duration  \
0                       4            0.159         330334   
1                       3            0.206         319149   
2                       4            0.310         310822   
3                       3            0.127         226733   
4                       3            0.785         229769   

                                           cifra_url  \
0  https://www.cifras.com.br/cifra/gesil-amarante...   
1  https://www.cifras.com.br/cifra/may-soares/nao...   
2  https://www.cifras.com.br/cifra/luciana-gomes/...   
3  https://www.cifras.com.br/cifra/gabriel-iglesi...   
4  https://www.cifras.com.br/cifra/indiegentes/eu...   

                                         spotify_url  
0  https://open.spotify.com/track/5twnhNJdliNQNJ0...  
1  https://open.spotify.com/track/6eeJsfPD88u7eNP...  
2  https://open.spotify.com/track/1PvicL958FjFDNE...  
3  https://open.spotify.com/track/0LPkPmpdjniUMKW...  
4  https://open.spotify.com/track/6sKhZ0D5zqqPlZo...  

[5 rows x 23 columns]

## First iteration of cifra data scraping

- Here, there is some information from Spotify, but the API blocked my scraper.
- Keep in mind that Spotify allow users to do this through their official [API](https://developer.spotify.com/documentation/web-api), only requiring an authorization and access token.
- In total, there is 4063 data points that include Spotify information.

In [5]:
cifras_df2.describe(include=[object])

url_artist url_song                              cifra_file_loc  \
count       59376    59376                                       59376   
unique      17779    46947                                       59376   
top     catolicas  saudade  cifras2/shirley-carvalhaes/eu-sou-deus.txt   
freq          540       55                                           1   

       cifra_key  
count      38185  
unique        74  
top            G  
freq        6315

In [6]:
cifras_df2.head()

url_artist               url_song  \
0  shirley carvalhaes            eu sou deus   
1      marcelo aquino     padrao pra eu amar   
2      marcelo aquino                 perola   
3      marcelo aquino  teus olhos nao mentem   
4  raphael de almeida        digno es senhor   

                                     cifra_file_loc cifra_key  
0        cifras2/shirley-carvalhaes/eu-sou-deus.txt       F#m  
1     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt        Eb  
2                 cifras2/marcelo-aquino/perola.txt         F  
3  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt        Eb  
4    cifras2/raphael-de-almeida/digno-es-senhor.txt         G

## Second iteration of cifra data scraping

- After removing from the scraping loop the acquisition of data from Spotify, I could gather about 60k data points.
- Note that I only saved in the csv file the location of the cifras organized by artist folders and the cifra key.
- I missed the chance to save the artist and song name from the website which would be more precise than simply saving the information from the URL. Since Portuguese has accentuation, the URL information does not bring that.
- Also, we may noticed that there are 74 cifra keys in the whole dataset, which suggests that there is some miss information there.

In [7]:
cifras_df3.describe(include=[object])

cifra_url    url_artist_name  \
count                                                5558               5558   
unique                                               5558               1599   
top     https://www.cifras.com.br/cifra/pura-inocencia...  hinos-avulsos-ccb   
freq                                                    1                111   

              artist_name artist_genre                  url_song_name  \
count                5558         5558                           5558   
unique               1599           48                           5090   
top     hinos avulsos ccb    sertanejo  se-todos-fossem-iguais-a-voce   
freq                  111         1436                              5   

                            song_name  \
count                            5558   
unique                           5090   
top     se todos fossem iguais a voce   
freq                                5   

                                     cifra_file_loc cifra_key  
count                                          5558      4416  
unique                                         5558        39  
top     cifras2/pura-inocencia/conheco-seu-fogo.txt         G  
freq                                              1       606

In [8]:
cifras_df3.head()

cifra_url     url_artist_name  \
0  https://www.cifras.com.br/cifra/pura-inocencia...      pura-inocencia   
1  https://www.cifras.com.br/cifra/nelsinho-rodri...  nelsinho-rodrigues   
2  https://www.cifras.com.br/cifra/labibe/um-dia-...              labibe   
3  https://www.cifras.com.br/cifra/celso-lira/ser...          celso-lira   
4  https://www.cifras.com.br/cifra/looney-tunes/p...        looney-tunes   

          artist_name    artist_genre     url_song_name         song_name  \
0      pura inocencia  samba-e-pagode  conheco-seu-fogo  conheco seu fogo   
1  nelsinho rodrigues          brasil    briga-de-casal    briga de casal   
2              labibe   nao-informada      um-dia-livre      um dia livre   
3          celso lira   nao-informada      sertao-maior      sertao maior   
4        looney tunes   nao-informada   patolino-o-mago   patolino o mago   

                                  cifra_file_loc cifra_key  
0    cifras2/pura-inocencia/conheco-seu-fogo.txt         E  
1  cifras2/nelsinho-rodrigues/briga-de-casal.txt         D  
2                cifras2/labibe/um-dia-livre.txt         E  
3            cifras2/celso-lira/sertao-maior.txt         E  
4       cifras2/looney-tunes/patolino-o-mago.txt         C

## Third iteration of cifra data scraping

- Here, I added another two columns for the URL but I didn't got the information from the website so I can get the accentuation.
- Also, I got information about artist genre and added in the dataset.
- In total, there are 5558 new data points added.

In [9]:
cifras_df4.describe(include=[object])

cifra_url    url_artist_name  \
count                                               21361              21361   
unique                                              21361               4271   
top     https://www.cifras.com.br/cifra/banda-musa-do-...  hinos-avulsos-ccb   
freq                                                    1                322   

              artist_name artist_genre url_song_name song_name  \
count               21361        21361         21361     21361   
unique               4344           48         17759     18145   
top     Hinos Avulsos Ccb    sertanejo      sem-voce  Sem Você   
freq                  289         4541            24        23   

                                           cifra_file_loc cifra_key  
count                                               21361     15629  
unique                                              21361       261  
top     cifras2/banda-musa-do-calypso/eu-quero-atitude...         G  
freq                                                    1      2199

In [10]:
cifras_df4.tail()

cifra_url  \
21356  https://www.cifras.com.br/cifra/the-seres/roda...   
21357  https://www.cifras.com.br/cifra/samprazer/sabo...   
21358  https://www.cifras.com.br/cifra/trio-da-huanna...   
21359  https://www.cifras.com.br/cifra/liu-e-leo/senh...   
21360  https://www.cifras.com.br/cifra/dino-franco-e-...   

            url_artist_name           artist_name    artist_genre  \
21356             the-seres             The Seres      rockn-roll   
21357             samprazer             Samprazer  samba-e-pagode   
21358        trio-da-huanna        Trio da Huanna           forro   
21359             liu-e-leo             Liu e Léo       sertanejo   
21360  dino-franco-e-mourai  Dino Franco e Mouraí       sertanejo   

             url_song_name            song_name  \
21356      roda-de-samsara      Roda de Samsara   
21357     sabor-de-hortela     Sabor de Hortelã   
21358   encontrei-um-jeito   Encontrei Um Jeito   
21359     senhora-da-penha     Senhora da Penha   
21360  lagrimas-das-flores  Lágrimas Das Flores   

                                          cifra_file_loc cifra_key  
21356              cifras2/the-seres/roda-de-samsara.txt         A  
21357             cifras2/samprazer/sabor-de-hortela.txt         C  
21358      cifras2/trio-da-huanna/encontrei-um-jeito.txt       NaN  
21359             cifras2/liu-e-leo/senhora-da-penha.txt         C  
21360  cifras2/dino-franco-e-mourai/lagrimas-das-flor...         B

## Last iteration of cifra data scraping

- Finally, I added the accentuation information to the artist and song names.
- This last version of data scrape acquire added a total of 21356 new data points.

In [11]:
len(cifras_df1)+len(cifras_df2)+len(cifras_df3)+len(cifras_df4)

90358

## Standardize

- In total, I gather a dataset with more than 90k data points.
- Let's take the last dataframe as our standard, meaning we need to adapt/remove some information such as artist/song name and Spotify information.

In [12]:
cifras_df4.columns

Index(['cifra_url', 'url_artist_name', 'artist_name', 'artist_genre',
       'url_song_name', 'song_name', 'cifra_file_loc', 'cifra_key'],
      dtype='object')

### Standardize Columns:
- **cifra_url**: website URL of the cifra, source of our scraping information.
- **url_artist_name**: artist's name as it appers in the URL and folder name.
- **artist_name**: artist's name in standard Portuguese. 
- **artist_genre**: artist music genre informed by another scraping process.
- **url_song_name**: song name as it appers in the URL and cifra file name.
- **song_name**: song's name in standard Portuguese. 
- **cifra_file_loc**: file location of the cifra
- **cifra_key**: cifra's music key in which the song is based.

## Artist Music Genre Information

- The csv file ```artist_genres.csv``` has a list that was scraped from cifras website pointing out artists to their music genre.

In [13]:
artist_genre_df = pd.read_csv("artist_genres.csv")
artist_genre_df.describe(include=[object])

genre  artist_name               artist_url
count           71468        71467                    71468
unique             89        71450                    71223
top     nao-informada  Terri Gibbs  phillips-craig-and-dean
freq            35120            2                        3

In [14]:
artist_genre_df.head()

genre        artist_name        artist_url
0  rockn-roll  "Dropkick Murphys  dropkick-murphys
1  rockn-roll               10cc              10cc
2  rockn-roll          12-Stones         12-stones
3  rockn-roll         1280 Almas        1280-almas
4  rockn-roll            16 Bits           16-bits

### Function to retrieve artist music genre based on ```url_artist_name```

In [15]:
def get_artist_genre(df, url_artist_name):
    return df[df["artist_url"] == url_artist_name]["genre"].to_string().split(" ")[-1]
print(get_artist_genre(artist_genre_df, "caetano-veloso"))

mpb


### Adaptations to ```cifras_df1```
- Replace the ```spotify_artist``` and ```spotify_song``` as ```artist_name``` and ```song_name```.
- Remove the other spotify information.
- Adapt ```url_artist``` and ```url_song``` to have "-" and change columns' name.
- Add ```artist_genre``` information.

In [16]:
cifras_df1.columns

Index(['url_artist', 'url_song', 'cifra_file_loc', 'cifra_key',
       'spotify_artist', 'spotify_song', 'spotify_album', 'spotify_popularity',
       'release_date', 'feature_danceability', 'feature_energy',
       'feature_acousticness', 'feature_instrumentalness', 'feature_key',
       'feature_liveness', 'feature_mode', 'feature_speechiness',
       'feature_tempo', 'feature_time_signature', 'feature_valence',
       'song_duration', 'cifra_url', 'spotify_url'],
      dtype='object')

In [17]:
# Remove spotify columns
cifras_df1 = cifras_df1.drop(columns=['spotify_album', 'spotify_popularity',
       'release_date', 'feature_danceability', 'feature_energy',
       'feature_acousticness', 'feature_instrumentalness', 'feature_key',
       'feature_liveness', 'feature_mode', 'feature_speechiness',
       'feature_tempo', 'feature_time_signature', 'feature_valence',
       'song_duration', 'spotify_url'])

# Rename columns
cifras_df1 = cifras_df1.rename(columns={"spotify_artist": "artist_name", "spotify_song": "song_name"})

cifras_df1.head()

url_artist                 url_song  \
0  gesil amarante jr           sou um milagre   
1         may soares         nao passes daqui   
2      luciana gomes  consequencia do sucesso   
3   gabriel iglesias               velho amor   
4        indiegentes     eu nao sou teu amigo   

                                     cifra_file_loc cifra_key  \
0       cifras/gesil-amarante-jr/sou-um-milagre.txt         F   
1            cifras/may-soares/nao-passes-daqui.txt         A   
2  cifras/luciana-gomes/consequencia-do-sucesso.txt         A   
3            cifras/gabriel-iglesias/velho-amor.txt         A   
4       cifras/indiegentes/eu-nao-sou-teu-amigo.txt        Em   

      spotify_artist             spotify_song  \
0  Gesil Amarante Jr           Sou um Milagre   
1         May Soares         Não Passes Daqui   
2      LUCIANA GOMES  Consequência do Sucesso   
3   Gabriel Iglesias               Velho Amor   
4        Indiegentes     Eu Não Sou Teu Amigo   

                                           cifra_url  
0  https://www.cifras.com.br/cifra/gesil-amarante...  
1  https://www.cifras.com.br/cifra/may-soares/nao...  
2  https://www.cifras.com.br/cifra/luciana-gomes/...  
3  https://www.cifras.com.br/cifra/gabriel-iglesi...  
4  https://www.cifras.com.br/cifra/indiegentes/eu...

In [18]:
new_col_url_artist = ["-".join(url_artist.split(" ")) for url_artist in cifras_df1['url_artist'].to_list()]
new_col_url_artist[0:4]

['gesil-amarante-jr', 'may-soares', 'luciana-gomes', 'gabriel-iglesias']

In [19]:
# Drop column 'url_artist' to replace with new_col_url_artist
cifras_df1 = cifras_df1.drop(columns=['url_artist'])
cifras_df1['url_artist_name'] = new_col_url_artist
cifras_df1.head()

url_song                                    cifra_file_loc  \
0           sou um milagre       cifras/gesil-amarante-jr/sou-um-milagre.txt   
1         nao passes daqui            cifras/may-soares/nao-passes-daqui.txt   
2  consequencia do sucesso  cifras/luciana-gomes/consequencia-do-sucesso.txt   
3               velho amor            cifras/gabriel-iglesias/velho-amor.txt   
4     eu nao sou teu amigo       cifras/indiegentes/eu-nao-sou-teu-amigo.txt   

  cifra_key     spotify_artist             spotify_song  \
0         F  Gesil Amarante Jr           Sou um Milagre   
1         A         May Soares         Não Passes Daqui   
2         A      LUCIANA GOMES  Consequência do Sucesso   
3         A   Gabriel Iglesias               Velho Amor   
4        Em        Indiegentes     Eu Não Sou Teu Amigo   

                                           cifra_url    url_artist_name  
0  https://www.cifras.com.br/cifra/gesil-amarante...  gesil-amarante-jr  
1  https://www.cifras.com.br/cifra/may-soares/nao...         may-soares  
2  https://www.cifras.com.br/cifra/luciana-gomes/...      luciana-gomes  
3  https://www.cifras.com.br/cifra/gabriel-iglesi...   gabriel-iglesias  
4  https://www.cifras.com.br/cifra/indiegentes/eu...        indiegentes

In [20]:
# Drop column 'url_song' to replace with new standard column
new_col_url_song = ["-".join(url_song.split(" ")) for url_song in cifras_df1['url_song'].to_list()]

cifras_df1 = cifras_df1.drop(columns=['url_song'])
cifras_df1['url_song_name'] = new_col_url_song
cifras_df1.head()

cifra_file_loc cifra_key  \
0       cifras/gesil-amarante-jr/sou-um-milagre.txt         F   
1            cifras/may-soares/nao-passes-daqui.txt         A   
2  cifras/luciana-gomes/consequencia-do-sucesso.txt         A   
3            cifras/gabriel-iglesias/velho-amor.txt         A   
4       cifras/indiegentes/eu-nao-sou-teu-amigo.txt        Em   

      spotify_artist             spotify_song  \
0  Gesil Amarante Jr           Sou um Milagre   
1         May Soares         Não Passes Daqui   
2      LUCIANA GOMES  Consequência do Sucesso   
3   Gabriel Iglesias               Velho Amor   
4        Indiegentes     Eu Não Sou Teu Amigo   

                                           cifra_url    url_artist_name  \
0  https://www.cifras.com.br/cifra/gesil-amarante...  gesil-amarante-jr   
1  https://www.cifras.com.br/cifra/may-soares/nao...         may-soares   
2  https://www.cifras.com.br/cifra/luciana-gomes/...      luciana-gomes   
3  https://www.cifras.com.br/cifra/gabriel-iglesi...   gabriel-iglesias   
4  https://www.cifras.com.br/cifra/indiegentes/eu...        indiegentes   

             url_song_name  
0           sou-um-milagre  
1         nao-passes-daqui  
2  consequencia-do-sucesso  
3               velho-amor  
4     eu-nao-sou-teu-amigo

In [30]:
new_column_genre = [get_artist_genre(artist_genre_df, artist)  for artist in cifras_df1['url_artist_name']]
new_column_genre[0:4]

['gospel', 'gospel', 'gospel', 'mpb']

In [31]:
cifras_df1['artist_genre'] = new_column_genre

In [32]:
cifras_df1.head()

cifra_file_loc cifra_key  \
0       cifras/gesil-amarante-jr/sou-um-milagre.txt         F   
1            cifras/may-soares/nao-passes-daqui.txt         A   
2  cifras/luciana-gomes/consequencia-do-sucesso.txt         A   
3            cifras/gabriel-iglesias/velho-amor.txt         A   
4       cifras/indiegentes/eu-nao-sou-teu-amigo.txt        Em   

      spotify_artist             spotify_song  \
0  Gesil Amarante Jr           Sou um Milagre   
1         May Soares         Não Passes Daqui   
2      LUCIANA GOMES  Consequência do Sucesso   
3   Gabriel Iglesias               Velho Amor   
4        Indiegentes     Eu Não Sou Teu Amigo   

                                           cifra_url    url_artist_name  \
0  https://www.cifras.com.br/cifra/gesil-amarante...  gesil-amarante-jr   
1  https://www.cifras.com.br/cifra/may-soares/nao...         may-soares   
2  https://www.cifras.com.br/cifra/luciana-gomes/...      luciana-gomes   
3  https://www.cifras.com.br/cifra/gabriel-iglesi...   gabriel-iglesias   
4  https://www.cifras.com.br/cifra/indiegentes/eu...        indiegentes   

             url_song_name artist_genre  
0           sou-um-milagre       gospel  
1         nao-passes-daqui       gospel  
2  consequencia-do-sucesso       gospel  
3               velho-amor          mpb  
4     eu-nao-sou-teu-amigo   indie-rock

### Adaptations to ```cifras_df1``` completed!

**Let's just reorganize the columns order to match the standard!**

In [37]:
cifras_df1 = cifras_df1[['cifra_url', 'url_artist_name', 'artist_name', 'artist_genre',
       'url_song_name', 'song_name', 'cifra_file_loc', 'cifra_key']]
cifras_df1.head()

cifra_url    url_artist_name  \
0  https://www.cifras.com.br/cifra/gesil-amarante...  gesil-amarante-jr   
1  https://www.cifras.com.br/cifra/may-soares/nao...         may-soares   
2  https://www.cifras.com.br/cifra/luciana-gomes/...      luciana-gomes   
3  https://www.cifras.com.br/cifra/gabriel-iglesi...   gabriel-iglesias   
4  https://www.cifras.com.br/cifra/indiegentes/eu...        indiegentes   

         artist_name artist_genre            url_song_name  \
0  Gesil Amarante Jr       gospel           sou-um-milagre   
1         May Soares       gospel         nao-passes-daqui   
2      LUCIANA GOMES       gospel  consequencia-do-sucesso   
3   Gabriel Iglesias          mpb               velho-amor   
4        Indiegentes   indie-rock     eu-nao-sou-teu-amigo   

                 song_name                                    cifra_file_loc  \
0           Sou um Milagre       cifras/gesil-amarante-jr/sou-um-milagre.txt   
1         Não Passes Daqui            cifras/may-soares/nao-passes-daqui.txt   
2  Consequência do Sucesso  cifras/luciana-gomes/consequencia-do-sucesso.txt   
3               Velho Amor            cifras/gabriel-iglesias/velho-amor.txt   
4     Eu Não Sou Teu Amigo       cifras/indiegentes/eu-nao-sou-teu-amigo.txt   

  cifra_key  
0         F  
1         A  
2         A  
3         A  
4        Em

### Adaptations to ```cifras_df2```
- Adapt ```url_artist``` and ```url_song``` to have "-" and change columns' name.
- Add ```artist_genre``` information.
- Add ```cifra_url```.
- Add ```artist_name``` and ```song_name``` according to the standard present before.

In [40]:
# New column for url_artist_name and url_song_name
new_col_url_artist = ["-".join(url_artist.split(" ")) for url_artist in cifras_df2['url_artist'].to_list()]
new_col_url_song = ["-".join(url_song.split(" ")) for url_song in cifras_df2['url_song'].to_list()]

# Add new columns
cifras_df2['url_artist_name'] = new_col_url_artist
cifras_df2['url_song_name'] = new_col_url_song

# Remove old version
cifras_df2 = cifras_df2.drop(columns=['url_artist', 'url_song'])

cifras_df2.head()

cifra_file_loc cifra_key  \
0        cifras2/shirley-carvalhaes/eu-sou-deus.txt       F#m   
1     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt        Eb   
2                 cifras2/marcelo-aquino/perola.txt         F   
3  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt        Eb   
4    cifras2/raphael-de-almeida/digno-es-senhor.txt         G   

      url_artist_name          url_song_name  
0  shirley-carvalhaes            eu-sou-deus  
1      marcelo-aquino     padrao-pra-eu-amar  
2      marcelo-aquino                 perola  
3      marcelo-aquino  teus-olhos-nao-mentem  
4  raphael-de-almeida        digno-es-senhor

In [41]:
new_column_genre = [get_artist_genre(artist_genre_df, artist)  for artist in cifras_df2['url_artist_name']]
new_column_genre[0:4]

['gospel', 'nao-informada', 'nao-informada', 'nao-informada']

In [42]:
cifras_df2['artist_genre'] = new_column_genre

In [43]:
cifras_df2.head()

cifra_file_loc cifra_key  \
0        cifras2/shirley-carvalhaes/eu-sou-deus.txt       F#m   
1     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt        Eb   
2                 cifras2/marcelo-aquino/perola.txt         F   
3  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt        Eb   
4    cifras2/raphael-de-almeida/digno-es-senhor.txt         G   

      url_artist_name          url_song_name   artist_genre  
0  shirley-carvalhaes            eu-sou-deus         gospel  
1      marcelo-aquino     padrao-pra-eu-amar  nao-informada  
2      marcelo-aquino                 perola  nao-informada  
3      marcelo-aquino  teus-olhos-nao-mentem  nao-informada  
4  raphael-de-almeida        digno-es-senhor         gospel

#### ```artist_name``` and ```song_name``` information present in log file

- Sadly, I did not save this information in the csv file, even though they were available and added to the log file.
- To solve this problem, let's try to retrieve the information from log file.

In [44]:
!head -n 30 cifras_scraper2.log

2024-01-14 00:02:15.052 DEBUG: 	 Folder to store cifras created named: cifras2
2024-01-14 00:02:17.913 INFO: 	 Number of song URLs available: 448204
2024-01-14 00:02:17.916 INFO: 	 Cifra URL #22226: https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus
2024-01-14 00:02:17.916 INFO: 	 	Artist: shirley carvalhaes
2024-01-14 00:02:17.916 INFO: 	 	Song: eu sou deus
2024-01-14 00:02:20.556 DEBUG: 	 	 Song web title: Eu Sou Deus - Shirley Carvalhaes, BR=True
2024-01-14 00:02:25.463 INFO: 	 	Cifra #1 retrieved! Key: F#m
2024-01-14 00:02:25.464 INFO: 	 	Cifra stored at: cifras2/shirley-carvalhaes/eu-sou-deus.txt
2024-01-14 00:02:25.464 DEBUG: 	 	Wait interval of 2.449136734008789 secs for cifra scrape.
2024-01-14 00:02:27.916 INFO: 	 Cifra URL #22227: https://www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar
2024-01-14 00:02:27.916 INFO: 	 	Artist: marcelo aquino
2024-01-14 00:02:27.916 INFO: 	 	Song: padrao pra eu amar
2024-01-14 00:02:29.699 DEBUG: 	 	 Song web title: Padrão

In [47]:
def search_log_file(filename):
    i = 0
    with open(filename, 'r') as f:
        for line in f:
            if "Cifra URL #" in line:
                print(line)
                i += 1
            if i > 10:
                break

search_log_file("cifras_scraper2.log")

2024-01-14 00:02:17.916 INFO: 	 Cifra URL #22226: https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus

2024-01-14 00:02:27.916 INFO: 	 Cifra URL #22227: https://www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar

2024-01-14 00:02:37.919 INFO: 	 Cifra URL #22228: https://www.cifras.com.br/cifra/marcelo-aquino/perola

2024-01-14 00:02:47.924 INFO: 	 Cifra URL #22229: https://www.cifras.com.br/cifra/marcelo-aquino/teus-olhos-nao-mentem

2024-01-14 00:02:57.924 INFO: 	 Cifra URL #22230: https://www.cifras.com.br/cifra/marcos-barrientos/dios-de-los-imposible

2024-01-14 00:02:59.662 INFO: 	 Cifra URL #22231: https://www.cifras.com.br/cifra/raphael-de-almeida/digno-es-senhor

2024-01-14 00:03:07.927 INFO: 	 Cifra URL #22232: https://www.cifras.com.br/cifra/raphael-de-almeida/amado-meu

2024-01-14 00:03:17.930 INFO: 	 Cifra URL #22233: https://www.cifras.com.br/cifra/laryssa-caroline/santo-toca-de-assis

2024-01-14 00:03:19.563 INFO: 	 Cifra URL #22234: https://www.cifras.co

In [48]:
def search_log_file(filename):
    i = 0
    new_info = None
    song_info = None
    with open(filename, 'r') as f:
        for line in f:
            if "Cifra URL #" in line:
                new_info = line.split("https://")[-1]
                print(new_info)
            if "Song web title: " in line:
                song_info = line.split("Song web title:")[-1]
                print(song_info)
            i += 1
            if i > 30:
                break

search_log_file("cifras_scraper2.log")

www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus

 Eu Sou Deus - Shirley Carvalhaes, BR=True

www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar

 Padrão Pra Eu Amar - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcelo-aquino/perola

 Pérola - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcelo-aquino/teus-olhos-nao-mentem

 Teus Olhos Não Mentem - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcos-barrientos/dios-de-los-imposible



In [49]:
!head -n 100 cifras_scraper2.log

2024-01-14 00:02:15.052 DEBUG: 	 Folder to store cifras created named: cifras2
2024-01-14 00:02:17.913 INFO: 	 Number of song URLs available: 448204
2024-01-14 00:02:17.916 INFO: 	 Cifra URL #22226: https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus
2024-01-14 00:02:17.916 INFO: 	 	Artist: shirley carvalhaes
2024-01-14 00:02:17.916 INFO: 	 	Song: eu sou deus
2024-01-14 00:02:20.556 DEBUG: 	 	 Song web title: Eu Sou Deus - Shirley Carvalhaes, BR=True
2024-01-14 00:02:25.463 INFO: 	 	Cifra #1 retrieved! Key: F#m
2024-01-14 00:02:25.464 INFO: 	 	Cifra stored at: cifras2/shirley-carvalhaes/eu-sou-deus.txt
2024-01-14 00:02:25.464 DEBUG: 	 	Wait interval of 2.449136734008789 secs for cifra scrape.
2024-01-14 00:02:27.916 INFO: 	 Cifra URL #22227: https://www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar
2024-01-14 00:02:27.916 INFO: 	 	Artist: marcelo aquino
2024-01-14 00:02:27.916 INFO: 	 	Song: padrao pra eu amar
2024-01-14 00:02:29.699 DEBUG: 	 	 Song web title: Padrão

In [50]:
def search_log_file(filename):
    i = 0
    stored_info = []
    new_stored_info = None
    new_info = None
    song_info = None
    with open(filename, 'r') as f:
        for line in f:
            if "Cifra URL #" in line:
                new_info = line.split("https://")[-1]
                print(new_info)
            if "Song web title: " in line:
                song_info = line.split("Song web title:")[-1]
                print(song_info)
                stored_info.append([new_info, song_info])
            i += 1
            if i > 30:
                break
        return stored_info

print(search_log_file("cifras_scraper2.log"))

www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus

 Eu Sou Deus - Shirley Carvalhaes, BR=True

www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar

 Padrão Pra Eu Amar - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcelo-aquino/perola

 Pérola - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcelo-aquino/teus-olhos-nao-mentem

 Teus Olhos Não Mentem - Marcelo Aquino, BR=True

www.cifras.com.br/cifra/marcos-barrientos/dios-de-los-imposible

[['www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus\n', ' Eu Sou Deus - Shirley Carvalhaes, BR=True\n'], ['www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar\n', ' Padrão Pra Eu Amar - Marcelo Aquino, BR=True\n'], ['www.cifras.com.br/cifra/marcelo-aquino/perola\n', ' Pérola - Marcelo Aquino, BR=True\n'], ['www.cifras.com.br/cifra/marcelo-aquino/teus-olhos-nao-mentem\n', ' Teus Olhos Não Mentem - Marcelo Aquino, BR=True\n']]


In [55]:
def search_log_file(filename):
    i = 0
    stored_info = []
    with open(filename, 'r') as f:
        for line in f:
            if "Cifra URL #" in line:
                new_info = "https://" + line.split("https://")[-1].split("\n")[0]
            if "Song web title: " in line:
                song_info = line.split("Song web title: ")[-1].split(", BR=")[0]
                stored_info.append([new_info, song_info])
            i += 1
            if i > 30:
                break
        return stored_info

print(search_log_file("cifras_scraper2.log"))

[['https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus', 'Eu Sou Deus - Shirley Carvalhaes'], ['https://www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar', 'Padrão Pra Eu Amar - Marcelo Aquino'], ['https://www.cifras.com.br/cifra/marcelo-aquino/perola', 'Pérola - Marcelo Aquino'], ['https://www.cifras.com.br/cifra/marcelo-aquino/teus-olhos-nao-mentem', 'Teus Olhos Não Mentem - Marcelo Aquino']]


In [59]:
def search_log_file(filename):
    i = 0
    stored_info = []
    with open(filename, 'r') as f:
        for line in f:
            print(line)
            if "Cifra URL #" in line:
                new_info = "https://" + line.split("https://")[-1].split("\n")[0]
            if "Song web title: " in line:
                info = line.split("Song web title: ")[-1].split(", BR=")[0]
                song_info, artist_info = info.split(" - ")
                stored_info.append([new_info, song_info, artist_info])
            i += 1
            if i > 100:
                break
        return stored_info

print(search_log_file("cifras_scraper2.log"))

2024-01-14 00:02:15.052 DEBUG: 	 Folder to store cifras created named: cifras2

2024-01-14 00:02:17.913 INFO: 	 Number of song URLs available: 448204

2024-01-14 00:02:17.916 INFO: 	 Cifra URL #22226: https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus

2024-01-14 00:02:17.916 INFO: 	 	Artist: shirley carvalhaes

2024-01-14 00:02:17.916 INFO: 	 	Song: eu sou deus

2024-01-14 00:02:20.556 DEBUG: 	 	 Song web title: Eu Sou Deus - Shirley Carvalhaes, BR=True

2024-01-14 00:02:25.463 INFO: 	 	Cifra #1 retrieved! Key: F#m

2024-01-14 00:02:25.464 INFO: 	 	Cifra stored at: cifras2/shirley-carvalhaes/eu-sou-deus.txt

2024-01-14 00:02:25.464 DEBUG: 	 	Wait interval of 2.449136734008789 secs for cifra scrape.

2024-01-14 00:02:27.916 INFO: 	 Cifra URL #22227: https://www.cifras.com.br/cifra/marcelo-aquino/padrao-pra-eu-amar

2024-01-14 00:02:27.916 INFO: 	 	Artist: marcelo aquino

2024-01-14 00:02:27.916 INFO: 	 	Song: padrao pra eu amar

2024-01-14 00:02:29.699 DEBUG: 	 	 Song web t

In [64]:
def search_log_file(filename):
    stored_info = []
    with open(filename, 'r') as f:
        for line in f:
            if "Cifra URL #" in line:
                new_info = "https://" + line.split("https://")[-1].split("\n")[0]
            if "Song web title: " in line:
                info = line.split("Song web title: ")[-1].split(", BR=")[0]
                if len(info.split(" - ")) == 2:
                    song_info, artist_info = info.split(" - ")
                else:
                    aux = info.split(" - ")
                    song_info = aux[0]
                    song_info = aux[1]
                
                stored_info.append([new_info, song_info, artist_info])
        return stored_info

df_log_file = pd.DataFrame(search_log_file("cifras_scraper2.log"), 
                           columns=["cifra_url","song_name","artist_name"])
df_log_file.head()

cifra_url              song_name  \
0  https://www.cifras.com.br/cifra/shirley-carval...            Eu Sou Deus   
1  https://www.cifras.com.br/cifra/marcelo-aquino...     Padrão Pra Eu Amar   
2  https://www.cifras.com.br/cifra/marcelo-aquino...                 Pérola   
3  https://www.cifras.com.br/cifra/marcelo-aquino...  Teus Olhos Não Mentem   
4  https://www.cifras.com.br/cifra/marcos-barrien...  Dios De Los Imposible   

          artist_name  
0  Shirley Carvalhaes  
1      Marcelo Aquino  
2      Marcelo Aquino  
3      Marcelo Aquino  
4   Marcos Barrientos

#### ```artist_name``` and ```song_name``` information got from the log file

- I could retrieve the information from the log file and store in a data frame.
- Now, we can get ```artist_name``` and ```song_name``` information from another dataframe similarly of how we got information about music genre.
- To do this, we need to generate ```cifra_url``` first and add to ```cifras_df2```.

In [65]:
def get_cifra_url(df):
    url = "https://www.cifras.com.br/cifra/"
    ret_list = [ url + "/".join(file_loc.split(".")[0].split("/")[1:]) for file_loc in df["cifra_file_loc"]]

    return ret_list
    
cifras_df2["cifra_url"] = get_cifra_url(cifras_df2)
cifras_df2.head()

cifra_file_loc cifra_key  \
0        cifras2/shirley-carvalhaes/eu-sou-deus.txt       F#m   
1     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt        Eb   
2                 cifras2/marcelo-aquino/perola.txt         F   
3  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt        Eb   
4    cifras2/raphael-de-almeida/digno-es-senhor.txt         G   

      url_artist_name          url_song_name   artist_genre  \
0  shirley-carvalhaes            eu-sou-deus         gospel   
1      marcelo-aquino     padrao-pra-eu-amar  nao-informada   
2      marcelo-aquino                 perola  nao-informada   
3      marcelo-aquino  teus-olhos-nao-mentem  nao-informada   
4  raphael-de-almeida        digno-es-senhor         gospel   

                                           cifra_url  
0  https://www.cifras.com.br/cifra/shirley-carval...  
1  https://www.cifras.com.br/cifra/marcelo-aquino...  
2  https://www.cifras.com.br/cifra/marcelo-aquino...  
3  https://www.cifras.com.br/cifra/marcelo-aquino...  
4  https://www.cifras.com.br/cifra/raphael-de-alm...

In [74]:
def get_artist_name_from_log(df_log_file, cifra_url):
    return df_log_file[df_log_file["cifra_url"] == cifra_url]["artist_name"].to_string().split("    ")[1:]
print(get_artist_name_from_log(df_log_file, "https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus"))

['Shirley Carvalhaes']


In [75]:
def get_song_name_from_log(df_log_file, cifra_url):
    return df_log_file[df_log_file["cifra_url"] == cifra_url]["song_name"].to_string().split("    ")[1:]
print(get_song_name_from_log(df_log_file, "https://www.cifras.com.br/cifra/shirley-carvalhaes/eu-sou-deus"))

['Eu Sou Deus']


In [76]:
new_column_artist_name = [get_artist_name_from_log(df_log_file, url)  for url in cifras_df2['cifra_url']]
new_column_song_name = [get_song_name_from_log(df_log_file, url)  for url in cifras_df2['cifra_url']]
new_column_song_name[0:4]

[['Eu Sou Deus'],
 ['Padrão Pra Eu Amar'],
 ['Pérola'],
 ['Teus Olhos Não Mentem']]

In [77]:
new_column_artist_name[0:4]

[['Shirley Carvalhaes'],
 ['Marcelo Aquino'],
 ['Marcelo Aquino'],
 ['Marcelo Aquino']]

In [95]:
cifras_df2['artist_name'] = [item for sublist in new_column_artist_name for item in sublist or [""]]
cifras_df2['song_name'] = [item for sublist in new_column_song_name for item in sublist or [""]]

In [96]:
cifras_df2.head()

cifra_file_loc cifra_key  \
0        cifras2/shirley-carvalhaes/eu-sou-deus.txt       F#m   
1     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt        Eb   
2                 cifras2/marcelo-aquino/perola.txt         F   
3  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt        Eb   
4    cifras2/raphael-de-almeida/digno-es-senhor.txt         G   

      url_artist_name          url_song_name   artist_genre  \
0  shirley-carvalhaes            eu-sou-deus         gospel   
1      marcelo-aquino     padrao-pra-eu-amar  nao-informada   
2      marcelo-aquino                 perola  nao-informada   
3      marcelo-aquino  teus-olhos-nao-mentem  nao-informada   
4  raphael-de-almeida        digno-es-senhor         gospel   

                                           cifra_url         artist_name  \
0  https://www.cifras.com.br/cifra/shirley-carval...  Shirley Carvalhaes   
1  https://www.cifras.com.br/cifra/marcelo-aquino...      Marcelo Aquino   
2  https://www.cifras.com.br/cifra/marcelo-aquino...      Marcelo Aquino   
3  https://www.cifras.com.br/cifra/marcelo-aquino...      Marcelo Aquino   
4  https://www.cifras.com.br/cifra/raphael-de-alm...  Raphael de Almeida   

               song_name  
0            Eu Sou Deus  
1     Padrão Pra Eu Amar  
2                 Pérola  
3  Teus Olhos Não Mentem  
4        Digno És Senhor

### Adaptations to ```cifras_df2``` completed!

**Let's just reorganize the columns order to match the standard!**

In [97]:
cifras_df2 = cifras_df2[['cifra_url', 'url_artist_name', 'artist_name', 'artist_genre',
       'url_song_name', 'song_name', 'cifra_file_loc', 'cifra_key']]
cifras_df2.head()

cifra_url     url_artist_name  \
0  https://www.cifras.com.br/cifra/shirley-carval...  shirley-carvalhaes   
1  https://www.cifras.com.br/cifra/marcelo-aquino...      marcelo-aquino   
2  https://www.cifras.com.br/cifra/marcelo-aquino...      marcelo-aquino   
3  https://www.cifras.com.br/cifra/marcelo-aquino...      marcelo-aquino   
4  https://www.cifras.com.br/cifra/raphael-de-alm...  raphael-de-almeida   

          artist_name   artist_genre          url_song_name  \
0  Shirley Carvalhaes         gospel            eu-sou-deus   
1      Marcelo Aquino  nao-informada     padrao-pra-eu-amar   
2      Marcelo Aquino  nao-informada                 perola   
3      Marcelo Aquino  nao-informada  teus-olhos-nao-mentem   
4  Raphael de Almeida         gospel        digno-es-senhor   

               song_name                                    cifra_file_loc  \
0            Eu Sou Deus        cifras2/shirley-carvalhaes/eu-sou-deus.txt   
1     Padrão Pra Eu Amar     cifras2/marcelo-aquino/padrao-pra-eu-amar.txt   
2                 Pérola                 cifras2/marcelo-aquino/perola.txt   
3  Teus Olhos Não Mentem  cifras2/marcelo-aquino/teus-olhos-nao-mentem.txt   
4        Digno És Senhor    cifras2/raphael-de-almeida/digno-es-senhor.txt   

  cifra_key  
0       F#m  
1        Eb  
2         F  
3        Eb  
4         G

### Adaptations to ```cifras_df3```
- Adapt ```artist_name``` and ```song_name``` according to log file which has written in standard Portuguese.

In [105]:
df_log_file3 = pd.DataFrame(search_log_file("cifras_scraper3.log"), 
                           columns=["cifra_url","song_name","artist_name"])
df_log_file3.head()

cifra_url          song_name  \
0  https://www.cifras.com.br/cifra/jayane/crianca...  Criança Sabe Tudo   
1  https://www.cifras.com.br/cifra/pulley/outside...    Outside Opinion   
2  https://www.cifras.com.br/cifra/chabelos/el-anexo           El Anexo   
3       https://www.cifras.com.br/cifra/spitz/spider             Spider   
4  https://www.cifras.com.br/cifra/andy-williams/...  Winter Wonderland   

     artist_name  
0         Jayane  
1         Pulley  
2       Chabelos  
3          Spitz  
4  Andy Williams

In [131]:
new_column_artist_name = [get_artist_name_from_log(df_log_file3, url)  for url in cifras_df3['cifra_url']]
new_column_song_name = [get_song_name_from_log(df_log_file3, url)  for url in cifras_df3['cifra_url']]

In [132]:
new_column_artist_name[0:100]

[[],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 [],
 ['Paulo Roberto'],
 ['Paulo Sérgio'],
 ['Chico César'],
 ['Lorena Chaves'],
 ['Alborada'],
 ['Geraldo Vandré'],
 ['Pirisca Grecco'],
 ['Medida Salvadora'],
 ['Alberto Moreno'],
 ['Alberto Moreno'],
 ['Evandro Mesquita'],
 ['Célio José'],
 ['Todd Snider'],
 ['Adryana Ribeiro'],
 ['Steven Wilson'],
 ['Caetano Veloso'],
 ['Zé Renato'],
 ['Erick e Leo'],
 ['Bruno Cardoso'],
 ['João Mineiro e Marciano'],
 ['Lorena Chaves'],
 ['Angélica'],
 ['Milton Nunes']]

**Since there are some initial cases that I could not find the information from the log file, I will simply maintain the information already present in the dataset even though it is not written in standard Portuguese.**

In [133]:
new_column_artist_name = [item[0] if item else cifras_df3["artist_name"].iloc[i] for i, item in enumerate(new_column_artist_name)]
new_column_artist_name[0:100]

['pura inocencia',
 'nelsinho rodrigues',
 'labibe',
 'celso lira',
 'looney tunes',
 'luciano maia',
 'nicolas pereira',
 'andre carvalho',
 'lucas felix',
 'amaia montero',
 'lucas felix',
 'thiago conrado',
 'amaia montero',
 'lucas felix',
 'os anonimos',
 'marcos paulo e rulian',
 'facundo cabral',
 'victor e leo',
 'hinos avulsos ccb',
 'fernando e sorocaba',
 'israel e rodolffo',
 'igo marques',
 'igo marques',
 'cosmo jarvis',
 'fernando e sorocaba',
 'raulzito e os panteras',
 'edgard scandurra',
 'rogerio ferrari',
 'alejandro balbis',
 'graveola e o lixo polifonico',
 'paulo coracao',
 'tomate',
 'amaia montero',
 'a dos velas',
 'igo marques',
 'o bando de maria',
 'hinos avulsos ccb',
 'paulo coracao',
 'o bando de maria',
 'o bando de maria',
 'zeca pagodinho',
 'dianne reeves',
 'igo marques',
 'no te va gustar',
 'paulo coracao',
 'brian cohen e sophia abrahao',
 'paulo coracao',
 'paulinho natureza',
 'eduardo costa',
 'jorge ben jor',
 'diabo na cruz',
 'wilson costa 

In [134]:
new_column_song_name = [item[0] if item else cifras_df3["song_name"].iloc[i] for i, item in enumerate(new_column_song_name)]
new_column_song_name[0:4]

['conheco seu fogo', 'briga de casal', 'um dia livre', 'sertao maior']

In [135]:
cifras_df3 = cifras_df3.drop(columns=['artist_name', 'song_name'])
cifras_df3['artist_name'] = new_column_artist_name
cifras_df3['song_name'] = new_column_song_name
cifras_df3.head()

cifra_url     url_artist_name  \
0  https://www.cifras.com.br/cifra/pura-inocencia...      pura-inocencia   
1  https://www.cifras.com.br/cifra/nelsinho-rodri...  nelsinho-rodrigues   
2  https://www.cifras.com.br/cifra/labibe/um-dia-...              labibe   
3  https://www.cifras.com.br/cifra/celso-lira/ser...          celso-lira   
4  https://www.cifras.com.br/cifra/looney-tunes/p...        looney-tunes   

     artist_genre     url_song_name  \
0  samba-e-pagode  conheco-seu-fogo   
1          brasil    briga-de-casal   
2   nao-informada      um-dia-livre   
3   nao-informada      sertao-maior   
4   nao-informada   patolino-o-mago   

                                  cifra_file_loc cifra_key  \
0    cifras2/pura-inocencia/conheco-seu-fogo.txt         E   
1  cifras2/nelsinho-rodrigues/briga-de-casal.txt         D   
2                cifras2/labibe/um-dia-livre.txt         E   
3            cifras2/celso-lira/sertao-maior.txt         E   
4       cifras2/looney-tunes/patolino-o-mago.txt         C   

          artist_name         song_name  
0      pura inocencia  conheco seu fogo  
1  nelsinho rodrigues    briga de casal  
2              labibe      um dia livre  
3          celso lira      sertao maior  
4        looney tunes   patolino o mago

### Adaptations to ```cifras_df3``` completed!

**Let's just reorganize the columns order to match the standard!**

In [140]:
cifras_df3 = cifras_df3[['cifra_url', 'url_artist_name', 'artist_name', 'artist_genre',
       'url_song_name', 'song_name', 'cifra_file_loc', 'cifra_key']]
cifras_df3.head()

cifra_url     url_artist_name  \
0  https://www.cifras.com.br/cifra/pura-inocencia...      pura-inocencia   
1  https://www.cifras.com.br/cifra/nelsinho-rodri...  nelsinho-rodrigues   
2  https://www.cifras.com.br/cifra/labibe/um-dia-...              labibe   
3  https://www.cifras.com.br/cifra/celso-lira/ser...          celso-lira   
4  https://www.cifras.com.br/cifra/looney-tunes/p...        looney-tunes   

          artist_name    artist_genre     url_song_name         song_name  \
0      pura inocencia  samba-e-pagode  conheco-seu-fogo  conheco seu fogo   
1  nelsinho rodrigues          brasil    briga-de-casal    briga de casal   
2              labibe   nao-informada      um-dia-livre      um dia livre   
3          celso lira   nao-informada      sertao-maior      sertao maior   
4        looney tunes   nao-informada   patolino-o-mago   patolino o mago   

                                  cifra_file_loc cifra_key  
0    cifras2/pura-inocencia/conheco-seu-fogo.txt         E  
1  cifras2/nelsinho-rodrigues/briga-de-casal.txt         D  
2                cifras2/labibe/um-dia-livre.txt         E  
3            cifras2/celso-lira/sertao-maior.txt         E  
4       cifras2/looney-tunes/patolino-o-mago.txt         C

In [141]:
cifras_df = pd.concat([cifras_df1, cifras_df2, cifras_df3, cifras_df4], ignore_index=True)
cifras_df.shape

(90358, 8)

In [142]:
cifras_df.to_csv("consolidate_cifras.csv")

## Dataset consolidated and saved!

In [143]:
cifras_df.describe(include=[object])

cifra_url    url_artist_name  \
count                                               90358              90358   
unique                                              90357              21856   
top     https://www.cifras.com.br/cifra/marcelo-aquino...  hinos-avulsos-ccb   
freq                                                    2                808   

              artist_name   artist_genre url_song_name song_name  \
count               90358          90358         90358     90358   
unique              22602             76         67020     68751   
top     Hinos Avulsos Ccb  nao-informada       saudade             
freq                  770          23135            84       213   

                                     cifra_file_loc cifra_key  
count                                         90358     61978  
unique                                        90358       281  
top     cifras/gesil-amarante-jr/sou-um-milagre.txt         G  
freq                                              1      9637